server.go

```diff
@@ -499,6 +530,19 @@ func (s *Server) Stop() {
 	close(s.stopChan)
 }
 
+// Restart will gracefully disconnect from AMQP exactly like `Stop` but instead
+// of returning from `ListenAndServe` it will set everything up again from
+// scratch and start listening again. This can be useful if a server restart is
+// wanted without running `ListenAndServe` in a loop.
+func (s *Server) Restart() {
+	// Restart is noop if not running.
+	if atomic.LoadInt32(&s.isRunning) == 0 {
+		return
+	}
+
+	s.restartChan <- struct{}{}
```

Comment:
I originally thought that using this when Acks fail would cause multiple failed Acks to potentially restart the server hundreds of times in a row. But I think that it won't actually do that. Instead I actually think that it will deadlock?

- We have 2 messages currently in memory at the server. Both will fail to Ack.
- None of them are Acked
- Msg1: Fails Ack
- Msg1: Call Restart()
- Msg1: Blocks
- Msg2: Fails Ack
- Msg2: Call Restart()
- Msg2: Blocks
-  `monitorAndWait()` consumes one restart
- Msg1: Unblocks
- Server graceful stop begins
- Server waits for all handlers to complete.
- Deadlock.

If we instead use a `select` where the default is just a no-op:

- We have 2 messages currently in memory at the server. Both will fail to Ack.
- None of them are Acked
- Msg1: Fails Ack
- Msg1: Call Restart()
- Msg1: **Won't Block**
- Msg2: Fails Ack
- Msg2: Call Restart()
- Msg2: **Won't Block**
-  `monitorAndWait()` consumes one restart
- Server graceful stop begins
- Server waits for all handlers to complete.
- **Any handlers calling Restart() will no-op because server gracefully waits for all handlers before reconnect**
- **No Deadlock**.

Reply:
Nice summary, thanks! I implemented your suggestion and manually confirmed this is working by inspecting the log messages while running the test.

```sh
2024/05/14 21:00:14 server: no listener on restartChan, ensure server is running
2024/05/14 21:00:14 server: no listener on restartChan, ensure server is running
2024/05/14 21:00:14 server: restarting server
```